packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -566,6 +594,57 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
     return Future<void>.value();
   }
 
+  final Map<DebugPreHotRestartCallback, String> _hotRestartCallbacks = <DebugPreHotRestartCallback, String>{};
+
+  /// Register a callback that will be invoked before a hot restart is called.
+  ///
+  /// In non-debug modes this method is a no-op. This can be used to release native
+  /// resources acquired through platform channels or `dart:ffi`. Future returning
+  /// callbacks will be awaited, allowing for async tear downs.
+  ///
+  /// The following sample code shows how to use debugRegisterHotRestartCallback to handle
+  /// tearing down a native resource acquired through `dart:ffi`. In this example, if
+  /// the `context` pointer is not passed through to the `_destroyContext` function before
+  /// a hot restart, the application will crash after a hot restart.
+  ///
+  /// ```dart
+  /// import 'dart:ffi';
+  /// import 'package:flutter/foundation.dart';
+  /// import 'package:flutter/widgets.dart';
+  ///
+  /// final DynamicLibrary _lib = DynamicLibrary.open('some_native_lib.dll');
+  ///
+  /// final Pointer<NativeType> Function() _createContext = _lib.lookupFunction<
+  ///     Pointer<NativeType> Function(),
+  ///     Pointer<NativeType> Function()>('Native_create');
+  ///
+  /// final void Function(Pointer<NativeType>) _destroyContext = _lib.lookupFunction<
+  ///     Void Function(Pointer<NativeType>),
+  ///     void Function(Pointer<NativeType>)>('Native_destroy');
+  ///
+  /// class NativeResourceService {
+  ///   NativeResourceService() {
+  ///     if (kDebugMode) {
+  ///       WidgetsBinding.instance.debugRegisterHotRestartCallback(
+  ///           () => _destroyContext(_context),
+  ///           debugLabel: 'NativeResourceService');
+  ///     }
+  ///   }
+  ///
+  ///   /// Acquire native resources that must be released before they can
+  ///   /// be re-acquired.
+  ///   late final Pointer<NativeType> _context = _createContext();
+  /// }
+  /// ```
```

Comment:
sample code should be done using the techniques described in https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#provide-sample-code

Reply:
Done 40df2518c2193be99ab2b7f5c07b9ac04615c198